File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head == None:
            return False
        elif head.next == None:
            return False
        slow = head.next
        fast = head.next.next
        while fast != None and slow != None:
            if fast == slow:
                return True
            else:
                fast = fast.next
                if fast == None:
                    return False
                else:
                    fast = fast.next
                slow = slow.next
        return False

File: Data Structures & Algorithms/palindrome-linked-list/submission-1.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func isPalindrome(_ head: ListNode?) -> Bool {
        if head == nil || head!.next == nil{
            return true
        } else if head!.next!.next == nil {
            return head!.val == head!.next!.val
        }
        var current = head!.next
        var next = current!.next
        while next!.next != nil {
            current = current!.next
            next = next!.next
        }
        current!.next = nil
        return head!.val == next!.val && self.isPalindrome(head!.next)
    }
}
